### ISSUE TITLE
Docstring linting

### ISSUE THREAD

**Issue by billbrod** [flatironinstitute/ccn-template#26]
Flake8 won't check docstrings, the most common way to do that is [pydocstyle](https://www.pydocstyle.org/en/stable/)

**Comment by billbrod** [comment]
Or [darglint](https://pypi.org/project/darglint/)?

**Comment by BalzaniEdoardo** [comment]
I have been testing out the two linters. I found pydocstyle more accurate in linting the numpy style docstrings. 
I'll provide the linting of the numpy function meshgrid as an example (the function is at the end of the comment).

### Linting `numpy.meshgrid` with` pydocstyle`
As expected I get no error messages
```terminal
(venv) (base) ebalzani$ pydocstyle test_linting.py --convention=numpy -v
['src/ccn_template/test_linting.py']
```
### Linting `numpy.meshgrid` with` darglint`
Here I get multiple violations instead.
```
(venv) (base) ebalzani$ darglint -s numpy --verbosity 2 src/ccn_template/test_linting.py 
test_linting.py:meshgrid:3: DAR401: Missing exception(s) in Raises section: -r ValueError
test_linting.py:meshgrid:13: DAR102: Excess parameter(s) in Docstring: + ...
test_linting.py:meshgrid:13: DAR102: Excess parameter(s) in Docstring: + x1
test_linting.py:meshgrid:13: DAR102: Excess parameter(s) in Docstring: + x2
test_linting.py:meshgrid:13: DAR102: Excess parameter(s) in Docstring: + xn
test_linting.py:meshgrid:13: DAR101: Missing parameter(s) in Docstring: - *xi
```

If you want to give it a try, below the function I used, my darglint is version 1.8.1 and pydocstyle is 6.3.0.

```python
def meshgrid(*xi, copy=True, sparse=False, indexing='xy'):
    """
    Return a list of coordinate matrices from coordinate vectors.

    Make N-D coordinate arrays for vectorized evaluations of
    N-D scalar/vector fields over N-D grids, given
    one-dimensional coordinate arrays x1, x2,..., xn.

    .. versionchanged:: 1.9
       1-D and 0-D cases are allowed.

    Parameters
    ----------
    x1, x2,..., xn : array_like
        1-D arrays representing the coordinates of a grid.
    indexing : {'xy', 'ij'}, optional
        Cartesian ('xy', default) or matrix ('ij') indexing of output.
        See Notes for more details.

        .. versionadded:: 1.7.0
    sparse : bool, optional
        If True the shape of the returned coordinate array for dimension *i*
        is reduced from ``(N1, ..., Ni, ... Nn)`` to
        ``(1, ..., 1, Ni, 1, ..., 1)``.  These sparse coordinate grids are
        intended to be use with :ref:`basics.broadcasting`.  When all
        coordinates are used in an expression, broadcasting still leads to a
        fully-dimensonal result array.

        Default is False.

        .. versionadded:: 1.7.0
    copy : bool, optional
        If False, a view into the original arrays are returned in order to
        conserve memory.  Default is True.  Please note that
        ``sparse=False, copy=False`` will likely return non-contiguous
        arrays.  Furthermore, more than one element of a broadcast array
        may refer to a single memory location.  If you need to write to the
        arrays, make copies first.

        .. versionadded:: 1.7.0

    Returns
    -------
    X1, X2,..., XN : list of ndarrays
        For vectors `x1`, `x2`,..., `xn` with lengths ``Ni=len(xi)``,
        returns ``(N1, N2, N3,..., Nn)`` shaped arrays if indexing='ij'
        or ``(N2, N1, N3,..., Nn)`` shaped arrays if indexing='xy'
        with the elements of `xi` repeated to fill the matrix along
        the first dimension for `x1`, the second for `x2` and so on.

    Notes
    -----
    This function supports both indexing conventions through the indexing
    keyword argument.  Giving the string 'ij' returns a meshgrid with
    matrix indexing, while 'xy' returns a meshgrid with Cartesian indexing.
    In the 2-D case with inputs of length M and N, the outputs are of shape
    (N, M) for 'xy' indexing and (M, N) for 'ij' indexing.  In the 3-D case
    with inputs of length M, N and P, outputs are of shape (N, M, P) for
    'xy' indexing and (M, N, P) for 'ij' indexing.  The difference is
    illustrated by the following code snippet::

        xv, yv = np.meshgrid(x, y, indexing='ij')
        for i in range(nx):
            for j in range(ny):
                # treat xv[i,j], yv[i,j]

        xv, yv = np.meshgrid(x, y, indexing='xy')
        for i in range(nx):
            for j in range(ny):
                # treat xv[j,i], yv[j,i]

    In the 1-D and 0-D case, the indexing and sparse keywords have no effect.

    See Also
    --------
    mgrid : Construct a multi-dimensional "meshgrid" using indexing notation.
    ogrid : Construct an open multi-dimensional "meshgrid" using indexing
            notation.
    how-to-index

    Examples
    --------
    >>> nx, ny = (3, 2)
    >>> x = np.linspace(0, 1, nx)
    >>> y = np.linspace(0, 1, ny)
    >>> xv, yv = np.meshgrid(x, y)
    >>> xv
    array([[0. , 0.5, 1. ],
           [0. , 0.5, 1. ]])
    >>> yv
    array([[0.,  0.,  0.],
           [1.,  1.,  1.]])

    The result of `meshgrid` is a coordinate grid:

    >>> import matplotlib.pyplot as plt
    >>> plt.plot(xv, yv, marker='o', color='k', linestyle='none')
    >>> plt.show()

    You can create sparse output arrays to save memory and computation time.

    >>> xv, yv = np.meshgrid(x, y, sparse=True)
    >>> xv
    array([[0. ,  0.5,  1. ]])
    >>> yv
    array([[0.],
           [1.]])

    `meshgrid` is very useful to evaluate functions on a grid. If the
    function depends on all coordinates, both dense and sparse outputs can be
    used.

    >>> x = np.linspace(-5, 5, 101)
    >>> y = np.linspace(-5, 5, 101)
    >>> # full coordinate arrays
    >>> xx, yy = np.meshgrid(x, y)
    >>> zz = np.sqrt(xx**2 + yy**2)
    >>> xx.shape, yy.shape, zz.shape
    ((101, 101), (101, 101), (101, 101))
    >>> # sparse coordinate arrays
    >>> xs, ys = np.meshgrid(x, y, sparse=True)
    >>> zs = np.sqrt(xs**2 + ys**2)
    >>> xs.shape, ys.shape, zs.shape
    ((1, 101), (101, 1), (101, 101))
    >>> np.array_equal(zz, zs)
    True

    >>> h = plt.contourf(x, y, zs)
    >>> plt.axis('scaled')
    >>> plt.colorbar()
    >>> plt.show()
    """
    ndim = len(xi)

    if indexing not in ['xy', 'ij']:
        raise ValueError(
            "Valid values for `indexing` are 'xy' and 'ij'.")

    s0 = (1,) * ndim
    output = [np.asanyarray(x).reshape(s0[:i] + (-1,) + s0[i + 1:])
              for i, x in enumerate(xi)]

    if indexing == 'xy' and ndim > 1:
        # switch first and second axis
        output[0].shape = (1, -1) + s0[2:]
        output[1].shape = (-1, 1) + s0[2:]

    if not sparse:
        # Return the full N-D matrix (not only the 1-D vector)
        output = np.broadcast_arrays(*output, subok=True)

    if copy:
        output = [x.copy() for x in output]

    return output
```

**Comment by BalzaniEdoardo** [comment]
this clarification comes from chatGPT, and I think it explains why darglint is confused by the *xi argument:

## pydocstyle

[pydocstyle](https://www.pydocstyle.org/en/stable/) is a static analysis tool for checking compliance with Python docstring conventions. It supports most of the well-known docstring formats such as those specified in PEP 257, numpy, and google style conventions.

pydocstyle primarily focuses on the style of the docstring, ensuring that they are properly formatted and contain necessary sections like "Args", "Returns", etc. However, it does not check if the docstrings' content accurately represents the code functionality. For example, if a function's docstring states that it takes in two parameters but the function actually takes in three, pydocstyle would not flag this discrepancy.

## darglint

[darglint](https://github.com/terrencepreilly/darglint) is another tool used to enforce docstring quality, but it takes a different approach compared to pydocstyle. Instead of focusing on the style and format, darglint ensures that the function's description matches its implementation.

This means darglint checks that the arguments, return types, and raised exceptions in the docstring match what the function actually does in the code. If you added or removed a parameter in the function but didn't update the docstring, darglint would give a warning or error.

In summary, while both tools are used to enhance the quality of docstrings, they focus on different aspects: pydocstyle checks the format and style of the docstrings, while darglint checks the accuracy of the docstrings against the actual code. You might use both tools together to get comprehensive coverage of docstring quality in your Python projects.

**Comment by BalzaniEdoardo** [comment]
I like the scope of **darglint**, but I would not include it as a mandatory check on the code yet because it can fail even when the conventions are respected and the code is fine (as in the meshgrid example above). Interesting tool however

**Comment by billbrod** [comment]
Also worth noting that `darglint` is no longer maintained (see [readme](https://github.com/terrencepreilly/darglint#project-status), repo has been archived, last update in Oct 2021)

**Comment by BalzaniEdoardo** [comment]
completed the note on docstrings and pydocstyle linting
